### PR TITLE
Api token v2/permissions for api token

### PR DESCRIPTION
**Pull request by innerdvations**
### What does it do?

- adds `custom` token type
- adds `permissions` array of actions to custom tokens
- adds `lastUsed` timestamp to api tokens
- api tokens authorization now checks custom token permissions
- api tokens authorization now updates `lastUsed` time when authorized

There are still a few areas with uncertainty, but I think it's worth merging in and fixing them later if necessary. Without this we're blocked in a few areas, since the regenerate feature is based off this branch to prevent conflicts. We can also integrate the front-end and QA can start e2e testing.

### Why is it needed?

To make API Tokens v2 work with custom permissions

### How to test it?

- Create/update/delete custom tokens using api-token service OR the admin api
- check if they work with the content api

```js
    // Create a token
    const payload = {
      type: 'custom',
      permissions: ['api::restaurant.restaurant.find'],
      name: 'custom token v2',
    };
    const token = await strapi.admin.services['api-token'].create(payload);

    // update a token
    const updatedToken = await strapi.admin.services['api-token'].update(token.id, {
      description: 'foobar',
      permissions: ['api::restaurant.restaurant.find', 'api::restaurant.restaurant.findOne'],
    });
```

Tests have been provided and existing tests updated:

```sh
# runs the authorization, controller, and service unit tests:
yarn test:unit api-token
# runs the e2e tests
node ./test/e2e.js api-token
```

### Related issue(s)/PR(s)

Front end for last used timestamp:
https://github.com/strapi/strapi/pull/14108

### CODE REVIEW

**Comment by Bassel17** [comment]
Some e2e tests failing

**Comment by innerdvations** [comment]
Which tests are failing? From what I can tell, the only ones that fail are the ones that also fail in the `features/api-token-v2 branch` (admin-role and related, seems to be caused by some expected user accounts not existing) but I may be missing something.

**Comment by Bassel17** [comment]
> Which tests are failing? From what I can tell, the only ones that fail are the ones that also fail in the `features/api-token-v2 branch` (admin-role and related, seems to be caused by some expected user accounts not existing) but I may be missing something.

oh yeah sorry about that, didn't notice they are the same

**Review by Bassel17 — APPROVED** [review]
LGTM

File: packages/core/admin/server/content-types/index.js [hunk 1]
```diff
@@ -5,4 +5,5 @@ module.exports = {
   user: { schema: require('./User') },
   role: { schema: require('./Role') },
   'api-token': { schema: require('./api-token') },
+  'token-permission': { schema: require('./token-permission') },
```
Comment (Convly) [inline packages/core/admin/server/content-types/index.js:8]: Just a thought but shouldn't we call this `api-token-permission` to keep consistency with `api-token`?

File: packages/core/admin/server/content-types/token-permission.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+
+module.exports = {
+  collectionName: 'strapi_token_permissions',
```
Comment (Convly) [inline packages/core/admin/server/content-types/token-permission.js:4]: Same as above

File: packages/core/admin/server/controllers/api-token.js [hunk 3]
```diff
@@ -85,6 +86,11 @@ module.exports = {
       attributes.description = trim(body.description);
     }
 
+    // Don't allow updating lastUsedAt time
+    if (has(attributes, 'lastUsedAt')) {
```
Comment (Convly) [inline packages/core/admin/server/controllers/api-token.js:90]: Isn't it already handled by the validator below (l94) ?

File: packages/core/admin/server/services/__tests__/api-token.test.js [hunk 4]
```diff
@@ -254,15 +288,209 @@ describe('API Token', () => {
...
+      },
+      entityService: {
+        load,
+      },
+    };
+
+    const res = await apiTokenService.update(id, updatedAttributes);
+
+    expect(deleteFn).toHaveBeenCalledTimes(1);
+    // expect(deleteFn).toHaveBeenCalledWith({
```
Comment (Convly) [inline packages/core/admin/server/services/__tests__/api-token.test.js:376]: Forgot to remove this?

File: packages/core/admin/server/services/api-token.js [hunk 5]
```diff
@@ -1,30 +1,106 @@
...
+  if (Object.keys(whereParams).length === 0) {
+    return null;
+  }
+
+  const token = await strapi
+    .query('admin::api-token')
+    .findOne({ select: SELECT_FIELDS, populate: POPULATE_FIELDS, where: whereParams });
+
+  if (!token) return token;
+  return flattenTokenPermissions(token);
```
Comment (Convly) [inline packages/core/admin/server/services/api-token.js:92]: Does it make sense to flatten everytime?
If a controller uses this call to get back a token + populate its permissions, then try to return this data as a basic content-api response, it'll fail to map the token-permission's attributes to the actual data.

Maybe we could decouple the flattening of the result from this method. What do you think?

File: packages/core/admin/server/services/api-token.js [hunk 6]
```diff
@@ -47,28 +125,60 @@ const hash = (accessKey) => {
...
+    );
+
+    const currentPermissions = await strapi.entityService.load(
+      'admin::api-token',
+      apiToken,
+      'permissions'
+    );
+
+    if (currentPermissions) {
+      Object.assign(result, { permissions: map('action', currentPermissions) });
```
Comment (Convly) [inline packages/core/admin/server/services/api-token.js:177]: Same as my previous comment, maybe we shouldn't flatten the permission objects when dealing with actual ApiToken entities

File: packages/core/admin/server/services/api-token.js [hunk 7]
```diff
@@ -92,25 +202,37 @@ For security reasons, prefer storing the secret in an environment variable and r
...
 const list = async () => {
-  return strapi.query('admin::api-token').findMany({
+  const tokens = await strapi.query('admin::api-token').findMany({
     select: SELECT_FIELDS,
+    populate: POPULATE_FIELDS,
     orderBy: { name: 'ASC' },
   });
+
+  if (!tokens) return tokens;
+  return tokens.map((token) => flattenTokenPermissions(token));
```
Comment (Convly) [inline packages/core/admin/server/services/api-token.js:217]: Same :point_up:

File: packages/core/admin/server/services/__tests__/api-token.test.js (hunk 4 above)
Comment (innerdvations) [reply packages/core/admin/server/services/__tests__/api-token.test.js:376]: Will discuss this issue later, lower priority performance enhancement issue

**Pull request merged**